### ISSUE TITLE
Add History Hooks

### ISSUE THREAD

**Issue by theAkito** [nushell/nushell#13376]
### Related problem

For example, when running a command, which starts with `git reset`, I do not want it in the shell history.

### Describe the solution you'd like

History hooks. For example, when you execute a command, starting with the string `git reset` a hook (essentially just a `def`) should be executed. The content of the `def` is allowed to be arbitrary.

For the related example problem, I could hook in a `def`, which deletes the `git reset` related command from the history.

### Describe alternatives you've considered

Manually doing everything. Obviously, not a good alternative.

### Additional context and details

Hooks are generic. They can be used for any advanced workflow tasks, related to what you execute inside the shell.

**Comment by NotTheDr01ds** [comment]
Keep in mind that simply starting the command with a leading space will prevent it from being added to the history.  However, I realize that if you forget to do this, then you do still have to clear it manually, and I get that it would be nice to have a hook for it.

If someone was to implement this specifically, perhaps it would be possible to extend `pre_execution` in some way that could prevent a particular command from being added to the history.  Perhaps a return record (since the return currently doesn't seem to be used) with `{ history: false }`?

E.g.:

```Nu
$env.config.hooks.pre_execution ++= {||
  if ((commandline) | str starts-with "git reset") {
    return { history: false }
  }
}
```

**Comment by theAkito** [comment]
> Keep in mind that simply starting the command with a leading space will prevent it from being added to the history.

Yes, I was actually part of the original discussion, when this was to be added. 😉 https://github.com/nushell/nushell/issues/6886

> However, I realize that if you forget to do this, then you do still have to clear it manually, and I get that it would be nice to have a hook for it.

I still stand by my points from the mentioned issue. I do not use the space thing, because I think, it is a bad solution. (Is there a configuration option to disable this "feature"?)



> If someone was to implement this specifically, perhaps it would be possible to extend `pre_execution` in some way that could prevent a particular command from being added to the history. Perhaps a return record (since the return currently doesn't seem to be used) with `{ history: false }`?

Could you please explain?

Do you mean, that the `def` would just need to return this record?

Otherwise, I do not know, how practical it would be to keep all those `if` inside `$env.config.hooks.pre_execution`. 🤔

**Comment by theAkito** [comment]
> * Let's start with clarifying some terminology - Do you really mean a `def` here?  Or a closure like the other hooks take?  If a `def`, can you give an example of how you see this working?

The hook would work as follows.

I register the command `git reset` in some kind of built-in Nushell configuration option. Internally, it evaluates, whether the currently run `commandline.startsWith("git reset")`.

If `true`, the closure or whatever code block is evaluated.
Inside that block, there is a custom command called, for example, `myhook1`.
`myhook1` is a `def` whereever and follows the same rules and principles as all other `def`s.

Essentially, the hook does not care about *what* is executed. It just executes a code block, no matter what is inside it.
`myhook1` in this example, deletes the last history item and this is what I implement manually.



> * Also, if you really just want to delete `git reset` commands from the history constantly, that could technically be done in the next `pre_prompt` hook, right?  I just think that solution would be slow on _each_ prompt, since history manipulation isn't the fastest.  I was thinking it would be better (i.e., more performant) to skip the history handling entirely by just checking the `commandline` and provide a way to suppress it from being added to the history.

That's a very good point. Yes, preventing it from being added would be great.

____

That said, my example for my use case is just one of many.
My idea for this issue was, to have a generic solution for any kind of `def`. It just happens, that in my case, I would like to delete the last item from the history. However, it could be any functionality.

I never dealt with `pre_prompt` manually before, so am not sure, whether this functionality already covers a more generic use case spectrum. Can it take a generic `def` or does it need special treatment?
